packages/go/analysis/ad/ad.go

```diff
@@ -387,37 +395,77 @@ func CalculateCrossProductBitmaps(firstNodeSet, secondNodeSet cardinality.Duplex
 	return resultEntities
 }
 
-func CalculateCrossProductNodeSets(firstNodeSet, secondNodeSet []*graph.Node, groupExpansions impact.PathAggregator) cardinality.Duplex[uint32] {
+func CalculateCrossProductBitmaps(groupExpansions impact.PathAggregator, nodeSets ...cardinality.Duplex[uint32]) cardinality.Duplex[uint32] {
+	if len(nodeSets) < 2 {
+		log.Errorf("cross products require at least 2 nodesets")
+		return cardinality.NewBitmap32()
+	}
+
 	//The intention is that the node sets being passed into this function contain all the first degree principals for control
 	var (
-		resultEntities  = cardinality.NewBitmap32()
-		firstSetUnroll  = cardinality.NewBitmap32()
-		secondSetUnroll = cardinality.NewBitmap32()
+		resultEntities   = cardinality.NewBitmap32()
+		unrollSet        = cardinality.NewBitmap32()
+		cardinalityCache = map[uint32]uint64{}
+		checkSet         = cardinality.NewBitmap32()
 	)
 
-	//Unroll all groups in our first nodeset into a bitmap
-	for _, entity := range firstNodeSet {
-		firstSetUnroll.Add(entity.ID.Uint32())
-		if entity.Kinds.ContainsOneOf(ad.Group) {
-			firstSetUnroll.Or(groupExpansions.Cardinality(entity.ID.Uint32()).(cardinality.Duplex[uint32]))
+	//Take the second of our node sets and unroll it all into a single bitmap
+	nodeSets[1].Each(func(id uint32) (bool, error) {
+		checkSet.Add(id)
+		idCardinality := groupExpansions.Cardinality(id)
+		idCardinalityCount := getCardinalityCount(id, idCardinality, cardinalityCache)
+		cardinalityCache[id] = idCardinalityCount
+		if idCardinalityCount > 0 {
+			cardinalityCache[id] = idCardinalityCount
```

Comment:
Line 417 seems like it might be redundant? `cardinalityCache[id] = idCardinalityCount`